### PR TITLE
Add environment to MkosiState 

### PR DESCRIPTION
**Pull request by pyfisch**
Add a environment attribute to MkosiState and initialize it from MkosiConfig. This change allows for the addition of derived variables like `IMAGE_ID` and `IMAGE_VERSION` to the environment passed to scripts.

Based on the PR #1134 by @mcassaniti and includes the requested changes by @DaanDeMeyer.
It uses dataclass [Post-init processing](https://docs.python.org/3/library/dataclasses.html#post-init-processing) to avoid ugly getters and setters.

### CODE REVIEW

File: mkosi/machine.py [hunk 1]
```diff
@@ -117,6 +117,10 @@ def build(self) -> None:
         if needs_build(self.config):
             check_root()
             check_native(self.config)
+
+            # Useful if testing within Docker
+            if parse_boolean(os.getenv("MKOSI_TEST_NO_NAMESPACE", "0")):
+                raise unittest.SkipTest("Build test skipped due to environment variable.")
```
Comment (DaanDeMeyer) [inline mkosi/machine.py:123]: Let's drop this and do it in a separate PR if you really want it included

File: mkosi/backend.py [hunk 2]
```diff
@@ -506,6 +506,7 @@ class MkosiConfig:
     skeleton_trees: List[Path]
     clean_package_metadata: Union[bool, str]
     remove_files: List[Path]
+    # Environment should not be used directly. Use MkosiState environment instead.
```
Comment (DaanDeMeyer) [inline mkosi/backend.py:509]: Can we drop this comment and replace it with a general comment above the config class itself to always use the respective field from MkosiState if it exists?

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]

File: mkosi/machine.py (hunk 1 above)
Comment (mcassaniti) [reply mkosi/machine.py:123]: See #1179

**Comment by DaanDeMeyer** [comment]
Note that we want perfect PRs in mkosi, so next time please use git rebase to fix the existing commits instead of adding new ones on top. I'll squash this PR this time but in the future please use git rebase.
